src/provider/model/Environment.ts

```diff
@@ -0,0 +1,74 @@
+import {injectable} from "inversify";
+import {Signal1, Signal2} from "../common/Signal";
+import {AsyncInit} from "../controller/AsyncInit";
+import {WindowEvent} from "openfin/_v2/api/events/base";
+import {Identity} from "openfin/_v2/main";
+import {Application} from "../../client/main";
+import {AppWindow} from "./AppWindow";
+
+export interface Environment {
+    windowCreated: Signal2<Identity, string>;
+    windowClosed: Signal1<Identity>;
+
+    /**
+     * Creates a new application, given an App Directory entry.
+     */
+    createApplication: (appInfo: Application) => Promise<AppWindow>;
+
+    /**
+     * Creates an `AppWindow` object for an existing window.
+     */
+    wrapApplication: (appInfo: Application, identity: Identity) => AppWindow;
+}
+
+@injectable()
+export class FinEnvironment extends AsyncInit implements Environment {
+    /**
+     * Indicates that a new window has been created.
+     * 
+     * When the service first starts, this signal will also be fired for any pre-existing windows.
+     * 
+     * Arguments: (identity: Identity, manifestUrl: string)
+     */
+    public readonly windowCreated: Signal2<Identity, string> = new Signal2();
+
+    /**
+     * Indicates that a window has been closed.
+     * 
+     * Arguments: (identity: Identity)
+     */
+    public readonly windowClosed: Signal1<Identity> = new Signal1();
+
+    public async createApplication(appInfo: Application): Promise<AppWindow> {
+        const app = await fin.Application.startFromManifest(appInfo.manifest);
+        return new AppWindow(app.identity, appInfo);
+    }
+
+    public wrapApplication(appInfo: Application, identity: Identity): AppWindow {
```

Comment:
This is currently set up so that `FinEnvironment` is the only place where `AppWindow` objects are actually instantiated. The intention here is that in a test environment we'll wire up the injector to plug a `TestEnvironment` instance into each `Environment` location, and that `TestEnvironment` will instead create some kind of `TestWindow` class that has the same interface as `AppWindow`, but doesn't use the `fin` API in any way.

To do that cleanly should really mean extracting an interface from `AppWindow`, similar to `FinEnvironment` and `Environment`, but kinda deferring the decision around naming and feasibility of that approach until we start creating some "real" unit tests based on this structure.

To cut down on boilerplate, may have the "test" versions of the environment and window classes _extend_ the real types and selectively override the necessary bits, rather than needing to fully re-implement everything (for instance, `AppWindow` may well have a lot of non-`fin` stuff in it, that we don't want to have to duplicate.

Basically... TBD. Still needs some more thought/discussion.

Reply:
Ok, think I've figured out why this is giving me a bad feeling
- We're exposing two different levels of abstraction with this one class - the signals exposing things at the`Identity`/`uuid` level, and the methods exposing things at the `AppWindow` level
- There's nothing about the `wrapApplication` method that means it needs to be in the same class as the signals, they more "just happen" to be grouped together
- We get this odd ping-pong back and forth between this and Model when responding to a `windowCreated` signal

So I think this should be either:
- We change the signals to send `AppWindow`s
- We change `createApplication` to return `Promise<Identity>`, and pull out `wrapApplication` into a factory elsewhere